geoportailv3/static/js/exclusionmanagerservice.js

```diff
@@ -0,0 +1,185 @@
+ * @return {boolean} Whether the array intersect or not.
+ * @private
+ */
+app.ExclusionManager.prototype.intersects_ = function(arr1, arr2) {
+  if (typeof arr1 == 'string') {
+    arr1 = /** @type {Array} */ (JSON.parse(arr1));
+  }
+  if (typeof arr2 == 'string') {
+    arr2 = /** @type {Array} */ (JSON.parse(arr2));
+  }
+  var concat = arr1.concat(arr2);
+  goog.array.sort(concat);
+  var i;
+  var len = concat.length;
+  for (i = 1; i < len; i++) {
+    if (concat[i] == concat[i - 1]) {
+      return true;
+    }
+  }
+  return false;
+};
+
+
+/**
+ * @param {ol.Map} map The OpenLayers Map.
+ * @param {ol.layer.Layer} layer1 The layer which was added or for which the
+ *     opacity property has changed.
+ * @private
+ */
+app.ExclusionManager.prototype.checkForLayerExclusion_ = function(map, layer1) {
+  var opacity = layer1.get('opacity');
+  // don't do anything if layer is not displayed on map
+  if (!goog.isNumber(opacity) || opacity === 0) {
+    return;
+  }
+
+  if (!goog.isDef(layer1.get('metadata'))) {
+    return;
+  }
+  var exclusion1 = layer1.get('metadata')['exclusion'];
+
+  if (!goog.isDef(exclusion1)) {
+    return;
+  }
+
+  var layers = map.getLayers().getArray();
+  var len = layers.length;
+  var i;
+  var layer2;
+  var msg;
+
+  var layersToHide = [];
+  for (i = 0; i < len; i++) {
+    layer2 = layers[i];
+    if (layer2 == layer1 || !goog.isDef(layer2.get('metadata')) ||
+        !goog.isDef(layer2.get('metadata')['exclusion'])) {
+      continue;
+    }
+
+    // check exclusion with current baselayer
+    var exclusion2 = layer2.get('metadata')['exclusion'];
+    opacity = layer2.get('opacity');
+    if (this.intersects_(exclusion1, exclusion2) && goog.isNumber(opacity) &&
+        opacity > 0) {
+      // layer to exclude is not the current base layer
+      if (i !== 0) {
+        layersToHide.push(layer2.get('label'));
+        layer2.set('opacity', 0);
+      } else {
+        this.backgroundLayerMgr_.set(map, this.blankLayer_);
+        msg = 'The background layer has been deactivated because the ' +
+            'layer <b>' + layer1.get('label') +
+            '</b> cannot be displayed on top of it';
+        this.notify_(msg);
+      }
+    }
+  }
+  if (layersToHide.length) {
+    msg = 'The layer(s) <b>' + layersToHide.join(',') +
+        '</b> have been deactivated because they cannot be displayed while ' +
+        ' the layer <b>' + layer1.get('label') +
+        '</b> is displayed';
+    this.notify_(msg);
+  }
+};
+
+
+/**
+ * @param {ol.Map} map The OpenLayers map.
+ */
+app.ExclusionManager.prototype.init = function(map) {
+  var layerOpacityListenerKeys = {};
+
+  // listen on layers being added to the map
+  // base layers switch should fire the event as well
+  map.getLayers().on('add', goog.bind(function(e) {
+    var layer = e.element;
+    this.checkForLayerExclusion_(map, layer);
+
+    // listen on opacity change
+    var key = goog.events.listen(layer, 'change:opacity', goog.bind(
```

Comment:
You don't need to use `goog.bind` with `goog.events.listen`. The fifth argument passed to `goog.events.listen` is the object used as `this` when calling the listener function.